horovod/keras/callbacks_impl.py

```diff
@@ -0,0 +1,177 @@
+        if self.params.get('steps'):
+            # The number of steps is provided in the parameters.
+            return self.params['steps']
+        elif self.params.get('samples') and self.params.get('batch_size'):
+            # Compute the number of steps per epoch using # of samples and a batch size.
+            return self.params['samples'] // self.params['batch_size']
+        else:
+            raise ValueError('Could not autodetect the number of steps per epoch. '
+                             'Please specify the steps_per_epoch parameter to the '
+                             '%s() or upgrade to the latest version of Keras.'
+                             % self.__class__.__name__)
+
+    def _adjust_learning_rate(self, epoch):
+        old_lr = self.backend.get_value(self.model.optimizer.lr)
+        new_lr = self.initial_lr * self.multiplier(epoch)
+        self.backend.set_value(self.model.optimizer.lr, new_lr)
+
+        if hasattr(self.model.optimizer, 'momentum') and self.momentum_correction:
+            # See the paper cited above for more information about momentum correction.
+            self.restore_momentum = self.backend.get_value(self.model.optimizer.momentum)
+            self.backend.set_value(self.model.optimizer.momentum,
+                        self.restore_momentum * new_lr / old_lr)
+
+    def _restore_momentum_if_needed(self):
+        if self.restore_momentum:
+            self.backend.set_value(self.model.optimizer.momentum, self.restore_momentum)
+            self.restore_momentum = None
+
+    def on_train_begin(self, logs=None):
+        self.initial_lr = self.backend.get_value(self.model.optimizer.lr)
+        if not self.staircase and not self.steps_per_epoch:
+            self.steps_per_epoch = self._autodetect_steps_per_epoch()
+
+    def on_epoch_begin(self, epoch, logs=None):
+        self.current_epoch = epoch
+
+    def on_batch_begin(self, batch, logs=None):
+        if (self.current_epoch < self.start_epoch or
+                (self.end_epoch is not None and self.current_epoch >= self.end_epoch)):
+            # Outside of the adjustment scope.
+            return
+
+        if self.staircase and batch == 0:
+            # Do on first batch of every epoch.
+            self._adjust_learning_rate(self.current_epoch)
+        elif not self.staircase:
+            epoch = self.current_epoch + float(batch) / self.steps_per_epoch
+            self._adjust_learning_rate(epoch)
+
+    def on_batch_end(self, batch, logs=None):
+        self._restore_momentum_if_needed()
+
+    def on_epoch_end(self, epoch, logs=None):
+        if logs is not None:
+            # Log current learning rate.
+            logs['lr'] = self.backend.get_value(self.model.optimizer.lr)
+
+
+class LearningRateWarmupCallbackImpl(LearningRateScheduleCallbackImpl):
+    def __init__(self, backend, warmup_epochs=5, momentum_correction=True, steps_per_epoch=None,
+                 verbose=0):
+        """
```

Comment:
Since you did not copy-paste other comments, we can nuke this one as well.

Reply:
Done.